src/design-pages/passwordless_gdm.rst

```diff
@@ -117,6 +117,32 @@ diagram.
   authentication mechanism. This will be explained separately for each
   mechanism.
 
+krb5_child
+****************
+``krb5_child`` is the helper binary in charge of Kerberos authentication. It
+follows the general model of separating the authencation in two steps:
+``preauthentication`` and ``authentication``. The first part opens a session
+to obtain the data related to the authentication (i.e. EIdP code). The
+``krb5_child`` keeps the state while the information is displayed to the user
+and they follow the necessary steps for authentication. At this point is when
+the status changes to ``authentication`` and proceeds with the authentication
+itself.
+
+This was a valid solution when SSSD was the one deciding which authentication
+method to use during the process. This is no longer the case, since with this
+new proposal it is the user who decides the mechanism to be used, so the
+current ``krb5_child`` design must be extended.
+
+During the ``preauthentication`` phase all authentication methods that are
+available to the user are checked and all necessary information (e.g. login
+URLs, codes, prompts) is obtained to proceed with the authentication.
+Since some authentication mechanisms need to keep a session open, these are
+marked as keep-alive. The information is displayed to the user and once
+the user enters the credentials, pam_sss switches to the ``authentication``
+phase and PAM responder serializes the credentials in the ``sss_auth_token``
+structure. ``krb5_child`` gets the authentication type and the credentials,
+and continues with the authentication process.
```

Comment:
Are you describing the current state or the future state here?

Reply:
Future state. Am I missing anything?

Reply:
"Since some authentication mechanisms need to keep a session open, these are
marked as keep-alive." Keep krb5_child alive for all methods to unify the behavior.

This is the sentence that confused me, since this is the current behavior - only selected methods keep krb5_child running. Now we want the "preauthentication" to obtain the list and all information about all methods, keep krb5_child alive and wait for response.

We could also rename "preauthentication" to "get prompts" or "list methods" to make its meaning more  clear.

Reply:
I've rewritten the sentence to make it clear that `krb5_child` is kept alive for all methods.

> We could also rename "preauthentication" to "get prompts" or "list methods" to make its meaning more clear.

I would prefer to avoid this change since this reference is used in more than ten files and I am making these changes outside of the main SSSD development branch, and will probably have to continue this way for a few more months. I've been working this way for about a year and a half now and it's bearable because my changes are fairly self-contained, but I've had a scare or two, and making this change would break that dynamic.